Adding contributor's guide (#16)
l3af-project/l3afd: CONTRIBUTING.md

# How To Contribute

## Some Ways to Contribute

> - Report potential bugs.

Reviewer: ```suggestion
- Report potential bugs
```

> - Suggest product enhancements.

Reviewer: ```suggestion
- Suggest product enhancements
```

> - Increase our test coverage.

Reviewer: ```suggestion
- Increase our test coverage
```

> - Fix a bug.

Reviewer: ```suggestion
- Fix a bug
```

> - Implement a requested enhancement.

Reviewer: ```suggestion
- Implement a requested enhancement
```

> - Improve our guides and documentation. L3af Guides, Docs, and api godoc are deployed from this repo.

Reviewer: ```suggestion
- Improve our guides and documentation
```

> 

Reviewer: ```suggestion
- Contribute code
```

> ## Clone and Provision Environment

Reviewer: I would probably move this section after the pull request guidelines

1. Make sure you have a GitHub account

> 2. Fork the l3afd repository to your GitHub user or organization. 

Reviewer: ```suggestion
2. Fork the l3afd repository to your GitHub user or organization
```

3. Clone your ${YOUR_GITHUB_USERNAME_OR_ORG}/l3afd fork into your GOPATH, and setup the base repository as upstream remote:

```
mkdir -p "${GOPATH}/src/github.com/l3af-project"
cd "${GOPATH}/src/github.com/l3af-project"
git clone https://github.com/${YOUR_GITHUB_USERNAME_OR_ORG}/l3afd.git
cd l3afd
git remote add upstream https://github.com/l3af-project/l3afd.git
```
4. Setup your [L3AF Development Environment on vagrant](https://github.com/l3af-project/l3af-arch/tree/main/dev_environment). 
5. Check the GitHub issues for [good tasks to get started](https://github.com/l3af-project/l3afd/issues). 
6. Start contributing.
7. Perform end-to-end test in dev environment.

## Pull Requests guidelines

> We actively welcome your pull requests.

Reviewer: ```suggestion
Great, you want to directly contribute to the l3af project and submit a pull request. It is recommended prior to working on a PR to submit an issue in github for the change you want to make describing the change and context around it. This gives the l3af team a chance to review the issue and provide feedback and work with you on the change. If you have any questions, please feel free to reach out to the l3af team via Slack or mail group. Below are some general guidelines to help ensure a successful PR approval. 
```

> 

Reviewer: ```suggestion
- Provide background why you are making the change and the issue it addresses
- List what is changing and provide a high-level summary of the change
- List any relevant dependencies linked to this change
- Describe the tests that you ran to verify your changes and add/update test cases
- Update relevant docs, especially if you've change API's
- Ensure all tests pass and that your code lints
- ```

> - If you've added code that should be tested, add tests.
> - If you've changed APIs, update the documentation.
> - Ensure the test suite passes.
> - Make sure your code lints. 

Reviewer: ```suggestion
```

# How To Contribute

> 

Reviewer 2: What are the Code of Conduct rules?  I would recommend https://www.contributor-covenant.org/version/2/0/code_of_conduct/

> 

Reviewer 2: Karan also pointed me to https://wiki.lfnetworking.org/display/L3AF/Getting+Started+with+L3AF which contains a bunch of information not in this document.  We should either incorporate it or point to that page.

## Some Ways to Contribute

> - Report potential bugs

Reviewer 2: The information on how to do this is missing from this document.  For example, there shouldn't be any need to clone the repo to report a bug (e.g., a documentation issue), or to suggest product enhancements.   There should instead be guidance around how to file a github issue and what the expectations are.   See https://github.com/microsoft/ebpf-for-windows/blob/master/CONTRIBUTING.md for an example (which we got from other projects, so we take no credit for that).

Furthermore, is there a security vulnerability reporting process?

- Suggest product enhancements
- Increase our test coverage
- Fix a bug
- Implement a requested enhancement
- Improve our guides and documentation
- Contribute code

## Clone and Provision Environment

1. Make sure you have a GitHub account
2. Fork the l3afd repository to your GitHub user or organization. 

> 3. Clone your ${YOUR_GITHUB_USERNAME_OR_ORG}/l3afd fork into your GOPATH, and setup the base repository as upstream remote:

Reviewer 2: ```suggestion
3. Clone your ${YOUR_GITHUB_USERNAME_OR_ORG}/l3afd fork into your GOPATH, and set up the base repository as upstream remote:
```

```
mkdir -p "${GOPATH}/src/github.com/l3af-project"
cd "${GOPATH}/src/github.com/l3af-project"
git clone https://github.com/${YOUR_GITHUB_USERNAME_OR_ORG}/l3afd.git
cd l3afd
git remote add upstream https://github.com/l3af-project/l3afd.git
```

> 4. Setup your [L3AF Development Environment on vagrant](https://github.com/l3af-project/l3af-arch/tree/main/dev_environment). 

Reviewer 2: ```suggestion
4. Set up your [L3AF Development Environment on vagrant](https://github.com/l3af-project/l3af-arch/tree/main/dev_environment). 
```
Is Vagrant really required for all contributions?  For example, if I'm contributing changes specific to Windows, why is that step required?

5. Check the GitHub issues for [good tasks to get started](https://github.com/l3af-project/l3afd/issues). 
6. Start contributing.
7. Perform end-to-end test in dev environment.

## Pull Requests guidelines

Great, you want to directly contribute to the l3af project and submit a pull request.
It is recommended prior to working on a PR to submit an issue in github for the change you want

> to make describing the change and context around it. This gives the l3af team a chance to review

Reviewer 2: ```suggestion
to make describing the change and context around it. This gives the l3af maintainers a chance to review
```
I would avoid using the word "team" in any documentation since it tends to imply you're talking about Walmart where instead this is an LF project so use open source terminology.

the issue and provide feedback and work with you on the change. If you have any questions, please
feel free to reach out to the l3af team via [Slack](https://app.slack.com/client/T02GD9YQJUT/C02GRTC0SAD) or
[mail](main@lists.l3af.io) group. Below are some general guidelines to help ensure a successful PR approval.

> 

Reviewer 2: Missing discussion of DCO

> 

Reviewer 2: Any coding conventions one is expected to follow?  (For comparison, see https://github.com/microsoft/ebpf-for-windows/blob/master/docs/DevelopmentGuide.md which we copied from other projects)

- Provide background why you are making the change and the issue it addresses
- List what is changing and provide a high-level summary of the change
- List any relevant dependencies linked to this change
- Describe the tests that you ran to verify your changes and add/update test cases
- Update relevant docs, especially if you've change API's

> - Ensure all tests pass and that your code lints

Reviewer 2: Need more info on what it means to "ensure that your code lints".   Shouldn't that be unnecessary to say, i.e., shouldn't it just be handled by the normal build process and also checked by github ci/cd?

# How To Contribute

## Some Ways to Contribute
- Report potential bugs
- Suggest product enhancements
- Increase our test coverage
- Fix a bug
- Implement a requested enhancement
- Improve our guides and documentation
- Contribute code

### Report potential bugs

First, **ensure the bug was not already reported** by searching on GitHub under
[Issues](https://github.com/l3af-project/l3afd/issues).

> If you found a related bug, you can help us by

Reviewer: ```suggestion
If you did not find a related bug, you can help us by
```

[submitting a GitHub Issue](https://github.com/l3af-project/l3afd/issues/new).

> The best bug reports provide a detailed description of the issue and step-by-step instructions

Reviewer: ```suggestion
A good bug report provides a detailed description of the issue and step-by-step instructions
```

for reliably reproducing the issue.

We will aim to triage issues in weekly TSC meetings. In case we are unable to repro the issue,

> we will request more information from you, the filer. There will be a waiting period of 2 weeks

Reviewer: ```suggestion
we will request more information from you. There will be a waiting period of 2 weeks
```

for the requested information and if there is no response, the issue will be closed. If this happens,

> please reopen the issue if you do get a repro and collect the requested information.

Reviewer: ```suggestion
please reopen the issue if you do get a repro and provide the requested information.
```

> However, in the best case, we would love it if you can submit a Pull Request with a fix.

Reviewer: ```suggestion
```

> ### Did you write a patch that fixes a bug?

Reviewer: ```suggestion
### Submitting a patch that fixes a bug
```

Fork the repo and make your changes.
Then open a new GitHub pull request with the patch.

* Ensure the PR description clearly describes the problem and solution.
  Include the relevant issue number if applicable.

> ## New Features

Reviewer: I would suggest moving this below "Pull Requests Guidelines"

You can request a new feature by [submitting a GitHub Issue](https://github.com/l3af-project/l3afd/issues/new).

If you would like to implement a new feature, please first
[submit a GitHub Issue](https://github.com/l3af-project/l3afd/issues/new) and
communicate your proposal so that the L3AF community can review and provide feedback. Getting
early feedback will help ensure your implementation work is accepted by the community.
This will also allow us to better coordinate our efforts and minimize duplicated effort.

> ## Pull Requests guidelines

Reviewer: ```suggestion
## Guidelines for Pull Requests
```

Great, you want to directly contribute to the l3af project and submit a pull request.
It is recommended prior to working on a PR to submit an issue in github for the change you want
to make describing the change and context around it. This gives the l3af maintainers a chance to review
the issue and provide feedback and work with you on the change. If you have any questions, please
feel free to reach out to the l3af team via [Slack](https://app.slack.com/client/T02GD9YQJUT/C02GRTC0SAD) or
[mail](main@lists.l3af.io) group. Below are some general guidelines to help ensure a successful PR approval.

- Provide background why you are making the change and the issue it addresses
- List what is changing and provide a high-level summary of the change
- List any relevant dependencies linked to this change
- Describe the tests that you ran to verify your changes and add/update test cases
- Update relevant docs, especially if you've change API's

> - Ensure all tests pass and that your code lints

Reviewer: ```suggestion
- Ensure all GitHub CI/CD checks pass
```
